### PR TITLE
Refactoring to abstract out calls to target dependent intrsincs

### PR DESCRIPTION
**Pull request by sumesh13**
- Introduce targetmachine features to carry target specific info. Currently lets different targets obtain their specific intrinsics 

Will move the AMD/NVPTX specific calls to a target specific backend files later. This is temporary placeholder before refactoring the backend files.

### CODE REVIEW

**Comment by ROCm-Apps-Test** [comment]
Can one of the admins verify this patch?

**Comment by ROCm-Apps-Test** [comment]
Can one of the admins verify this patch?

**Comment by whchung** [comment]
jenkins: test this please

File: tensorflow/compiler/xla/service/gpu/amdgpu_compiler.cc [hunk 1]
```diff
@@ -353,6 +354,10 @@ StatusOr<std::unique_ptr<Executable>> AMDGPUCompiler::RunBackend(
   llvm_module.setTargetTriple(kTargetTriple);
   llvm_module.setDataLayout(kDataLayout);
 
+
+  // COnstruct targetmachine features based on triple - base class that provides features of
```
Comment (whchung) [inline tensorflow/compiler/xla/service/gpu/amdgpu_compiler.cc]: typo?

File: tensorflow/compiler/xla/service/gpu/ir_emitter_nested.cc [hunk 2]
```diff
@@ -38,7 +38,7 @@ namespace gpu {
 IrEmitterNested::IrEmitterNested(const HloModuleConfig& hlo_module_config,
                                  const HloComputation& nested_computation,
                                  IrEmitterContext* ir_emitter_context)
-    : IrEmitter(hlo_module_config, ir_emitter_context, /*is_nested=*/true) {
+    : IrEmitter(hlo_module_config, ir_emitter_context, /*is_nested=*/true, NULL) {
```
Comment (whchung) [inline tensorflow/compiler/xla/service/gpu/ir_emitter_nested.cc]: `nullptr`

File: tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.h [hunk 3]
```diff
@@ -48,7 +49,8 @@ class IrEmitterUnnested : public IrEmitter {
  public:
   IrEmitterUnnested(const HloModuleConfig& hlo_module_config,
                     const HloComputation* hlo_computation,
-                    IrEmitterContext* ir_emitter_context);
+                    IrEmitterContext* ir_emitter_context, 
+                    TargetMachineFeatures* target_machine_feature=NULL);
```
Comment (whchung) [inline tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.h]: `nullptr`

File: tensorflow/stream_executor/rocm/rocm_driver.cc [hunk 4]
```diff
@@ -490,9 +490,8 @@ ROCMDriver::DeviceGetSharedMemConfig(int device_ordinal) {
   // FIXME - need to set a 32-bit value here
   unsigned char valueC = static_cast<unsigned char>(value);
   uint32_t value32 = (valueC << 24) | (valueC << 16) | (valueC << 8) | (valueC) ;
-  assert (value32 == value); // if mismatch this indicates case where hipMemsetAsyc can't emulate hipMemSetD32
```
Comment (whchung) [inline tensorflow/stream_executor/rocm/rocm_driver.cc]: @sumesh13 not so sure the rationale behind changes in this file, could you help clarify?

Comment (sumesh13) [reply tensorflow/stream_executor/rocm/rocm_driver.cc]: let me dig up the example to see why I did that ..

Comment (sumesh13) [reply tensorflow/stream_executor/rocm/rocm_driver.cc]: Let me bring this change in a different PR while I look for the test case

File: tensorflow/stream_executor/rocm/rocm_driver.cc [hunk 5]
```diff
@@ -491,8 +491,8 @@ ROCMDriver::DeviceGetSharedMemConfig(int device_ordinal) {
   unsigned char valueC = static_cast<unsigned char>(value);
   uint32_t value32 = (valueC << 24) | (valueC << 16) | (valueC << 8) | (valueC) ;
   assert (value32 == value); // if mismatch this indicates case where hipMemsetAsyc can't emulate hipMemSetD32
-  hipError_t res =
-      hipMemsetAsync(pointer, value, uint32_count*4, stream);
+ hipError_t res =
```
Comment (whchung) [inline tensorflow/stream_executor/rocm/rocm_driver.cc]: nitpicking: please fix indentation

Comment (sumesh13) [reply tensorflow/stream_executor/rocm/rocm_driver.cc]: hmm..Dont see it locally..let me redit the line..

**Pull request merged**